File: app/src/main/java/com/etesync/syncadapter/ui/importlocal/ImportActivity.java
package com.etesync.syncadapter.ui.importlocal;

import android.accounts.Account;
import android.app.Activity;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.etesync.syncadapter.R;
import com.etesync.syncadapter.model.CollectionInfo;
import com.etesync.syncadapter.ui.BaseActivity;

public class ImportActivity extends BaseActivity implements SelectImportMethod, ResultFragment.OnImportCallback, DialogInterface {
    public final static String EXTRA_ACCOUNT = "account",
            EXTRA_COLLECTION_INFO = "collectionInfo";

    private Account account;
    protected CollectionInfo info;

    public static Intent newIntent(Context context, Account account, CollectionInfo info) {
        Intent intent = new Intent(context, ImportActivity.class);
        intent.putExtra(ImportActivity.EXTRA_ACCOUNT, account);
        intent.putExtra(ImportActivity.EXTRA_COLLECTION_INFO, info);
        return intent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        setTitle(getString(R.string.import_dialog_title));

        account = getIntent().getExtras().getParcelable(EXTRA_ACCOUNT);
        info = (CollectionInfo) getIntent().getExtras().getSerializable(EXTRA_COLLECTION_INFO);

        if (savedInstanceState == null)
            getSupportFragmentManager().beginTransaction()
                    .add(android.R.id.content, new ImportActivity.SelectImportFragment())
                    .commit();
    }

    @Override
    public void importFile() {
        getSupportFragmentManager().beginTransaction()
                .add(ImportFragment.newInstance(account, info), null)
                .commit();

    }

    @Override
    public void importAccount() {
        if (info.type == CollectionInfo.Type.CALENDAR) {
            getSupportFragmentManager().beginTransaction()
                    .replace(android.R.id.content,
                            LocalCalendarImportFragment.newInstance(account, info))
                    .addToBackStack(LocalCalendarImportFragment.class.getName())
                    .commit();
        } else if (info.type == CollectionInfo.Type.ADDRESS_BOOK) {
            getSupportFragmentManager().beginTransaction()
                    .replace(android.R.id.content,
                            LocalContactImportFragment.newInstance(account, info))
                    .addToBackStack(LocalContactImportFragment.class.getName())
                    .commit();
        }
        setTitle(getString(R.string.import_select_account));
    }

    private void popBackStack() {
        if (!getSupportFragmentManager().popBackStackImmediate()) {
            finish();
        } else {
            setTitle(getString(R.string.import_dialog_title));
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            popBackStack();
            return true;
        }
        return false;
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            popBackStack();
            return true;
        }
        return super.onKeyDown(keyCode, event);
    }

    @Override
    public void onImportResult(ResultFragment.ImportResult importResult) {
        ResultFragment fragment = ResultFragment.newInstance(importResult);
        fragment.show(getSupportFragmentManager(), "importResult");
    }

    @Override
    public void cancel() {
        finish();
    }

    @Override
    public void dismiss() {
        finish();
    }


    public static class SelectImportFragment extends Fragment {

        private SelectImportMethod mSelectImportMethod;

        @Override
        public void onAttach(Context context) {
            super.onAttach(context);
            // This makes sure that the container activity has implemented
            // the callback interface. If not, it throws an exception
            try {
                mSelectImportMethod = (SelectImportMethod) getActivity();
            } catch (ClassCastException e) {
                throw new ClassCastException(getActivity().toString()
                        + " must implement MyInterface ");
            }
        }

        @SuppressWarnings("deprecation")
        @Override
        public void onAttach(Activity activity) {
            super.onAttach(activity);
            // This makes sure that the container activity has implemented
            // the callback interface. If not, it throws an exception
            try {
                mSelectImportMethod = (SelectImportMethod) activity;
            } catch (ClassCastException e) {
                throw new ClassCastException(activity.toString()
                        + " must implement MyInterface ");
            }
        }

        public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
            View v = inflater.inflate(R.layout.import_actions_list, container, false);

            View card = v.findViewById(R.id.import_file);
            ImageView img = (ImageView) card.findViewById(R.id.action_icon);
            TextView text = (TextView) card.findViewById(R.id.action_text);
            img.setImageResource(R.drawable.ic_file_white);
            text.setText(R.string.import_button_file);
            card.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View aView) {
                    mSelectImportMethod.importFile();
                }
            });

            card = v.findViewById(R.id.import_account);
            img = (ImageView) card.findViewById(R.id.action_icon);
            text = (TextView) card.findViewById(R.id.action_text);
            img.setImageResource(R.drawable.ic_account_circle_white);
            text.setText(R.string.import_button_local);
            card.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View aView) {
                    mSelectImportMethod.importAccount();
                }
            });

            return v;
        }
    }
}

File: app/src/main/java/com/etesync/syncadapter/ui/importlocal/ImportActivity.kt
package com.etesync.syncadapter.ui.importlocal

import android.accounts.Account
import android.app.Activity
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.os.Bundle
import android.view.*
import android.widget.ImageView
import android.widget.TextView
import androidx.fragment.app.Fragment
import com.etesync.syncadapter.R
import com.etesync.syncadapter.model.CollectionInfo
import com.etesync.syncadapter.ui.BaseActivity

class ImportActivity : BaseActivity(), SelectImportMethod, ResultFragment.OnImportCallback, DialogInterface {

    private lateinit var account: Account
    protected lateinit var info: CollectionInfo

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        supportActionBar!!.setDisplayHomeAsUpEnabled(true)

        title = getString(R.string.import_dialog_title)

        account = intent.extras!!.getParcelable(EXTRA_ACCOUNT)!!
        info = intent.extras!!.getSerializable(EXTRA_COLLECTION_INFO) as CollectionInfo

        if (savedInstanceState == null)
            supportFragmentManager.beginTransaction()
                    .add(android.R.id.content, ImportActivity.SelectImportFragment())
                    .commit()
    }

    override fun importFile() {
        supportFragmentManager.beginTransaction()
                .add(ImportFragment.newInstance(account, info), null)
                .commit()

    }

    override fun importAccount() {
        if (info.type == CollectionInfo.Type.CALENDAR) {
            supportFragmentManager.beginTransaction()
                    .replace(android.R.id.content,
                            LocalCalendarImportFragment.newInstance(account, info))
                    .addToBackStack(LocalCalendarImportFragment::class.java.name)
                    .commit()
        } else if (info.type == CollectionInfo.Type.ADDRESS_BOOK) {
            supportFragmentManager.beginTransaction()
                    .replace(android.R.id.content,
                            LocalContactImportFragment.newInstance(account, info))
                    .addToBackStack(LocalContactImportFragment::class.java.name)
                    .commit()
        }
        title = getString(R.string.import_select_account)
    }

    private fun popBackStack() {
        if (!supportFragmentManager.popBackStackImmediate()) {
            finish()
        } else {
            title = getString(R.string.import_dialog_title)
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            popBackStack()
            return true
        }
        return false
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean {
        if (keyCode == KeyEvent.KEYCODE_BACK) {
            popBackStack()
            return true
        }
        return super.onKeyDown(keyCode, event)
    }

    override fun onImportResult(importResult: ResultFragment.ImportResult) {
        val fragment = ResultFragment.newInstance(importResult)
        supportFragmentManager.beginTransaction()
                .add(fragment, "importResult")
                .commitAllowingStateLoss()
    }

    override fun cancel() {
        finish()
    }

    override fun dismiss() {
        finish()
    }


    class SelectImportFragment : Fragment() {

        private var mSelectImportMethod: SelectImportMethod? = null

        override fun onAttach(context: Context?) {
            super.onAttach(context)
            // This makes sure that the container activity has implemented
            // the callback interface. If not, it throws an exception
            try {
                mSelectImportMethod = activity as SelectImportMethod?
            } catch (e: ClassCastException) {
                throw ClassCastException(activity!!.toString() + " must implement MyInterface ")
            }

        }

        override fun onAttach(activity: Activity?) {
            super.onAttach(activity)
            // This makes sure that the container activity has implemented
            // the callback interface. If not, it throws an exception
            try {
                mSelectImportMethod = activity as SelectImportMethod?
            } catch (e: ClassCastException) {
                throw ClassCastException(activity!!.toString() + " must implement MyInterface ")
            }

        }

        override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
            val v = inflater.inflate(R.layout.import_actions_list, container, false)

            var card = v.findViewById<View>(R.id.import_file)
            var img = card.findViewById<View>(R.id.action_icon) as ImageView
            var text = card.findViewById<View>(R.id.action_text) as TextView
            img.setImageResource(R.drawable.ic_file_white)
            text.setText(R.string.import_button_file)
            card.setOnClickListener { mSelectImportMethod!!.importFile() }

            card = v.findViewById(R.id.import_account)
            img = card.findViewById<View>(R.id.action_icon) as ImageView
            text = card.findViewById<View>(R.id.action_text) as TextView
            img.setImageResource(R.drawable.ic_account_circle_white)
            text.setText(R.string.import_button_local)
            card.setOnClickListener { mSelectImportMethod!!.importAccount() }

            if ((activity as ImportActivity).info.type == CollectionInfo.Type.TASKS) {
                card.visibility = View.GONE
            }

            return v
        }
    }

    companion object {
        val EXTRA_ACCOUNT = "account"
        val EXTRA_COLLECTION_INFO = "collectionInfo"

        fun newIntent(context: Context, account: Account, info: CollectionInfo): Intent {
            val intent = Intent(context, ImportActivity::class.java)
            intent.putExtra(ImportActivity.EXTRA_ACCOUNT, account)
            intent.putExtra(ImportActivity.EXTRA_COLLECTION_INFO, info)
            return intent
        }
    }
}
